// @node-wot/binding-http/dist/codecs/tuya-codec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class HttpTuyaCodec {
    getMediaType() {
        return "application/json+tuya";
    }
    bytesToValue(bytes, schema, parameters) {
        const parsedBody = JSON.parse(bytes.toString());
        if (!parsedBody.success)
            throw new Error(parsedBody.msg ? parsedBody.msg : JSON.stringify(parsedBody));
        for (const key in parsedBody.result) {
            if (parsedBody.result[key].code === schema["tuya:PropertyName"]) {
                return parsedBody.result[key].value;
            }
        }
        throw new Error("Property not found");
    }
    valueToBytes(value, schema, parameters) {
        const obj = {
            commands: [
                {
                    code: schema["tuya:PropertyName"],
                    value: value,
                },
            ],
        };
        const body = JSON.stringify(obj);
        return Buffer.from(body);
    }
}
exports.default = HttpTuyaCodec;
