// @node-wot/core/dist/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLoggers = exports.createErrorLogger = exports.createWarnLogger = exports.createDebugLogger = exports.createInfoLogger = void 0;
const debug_1 = __importDefault(require("debug"));
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["info"] = 0] = "info";
    LogLevel[LogLevel["debug"] = 1] = "debug";
    LogLevel[LogLevel["warn"] = 2] = "warn";
    LogLevel[LogLevel["error"] = 3] = "error";
})(LogLevel || (LogLevel = {}));
function createLogger(namespaces, logLevel) {
    const namespace = ["node-wot", ...namespaces, LogLevel[logLevel]].join(":");
    return (0, debug_1.default)(namespace);
}
function createInfoLogger(...namespaces) {
    return createLogger(namespaces, LogLevel.info);
}
exports.createInfoLogger = createInfoLogger;
function createDebugLogger(...namespaces) {
    return createLogger(namespaces, LogLevel.debug);
}
exports.createDebugLogger = createDebugLogger;
function createWarnLogger(...namespaces) {
    return createLogger(namespaces, LogLevel.warn);
}
exports.createWarnLogger = createWarnLogger;
function createErrorLogger(...namespaces) {
    return createLogger(namespaces, LogLevel.error);
}
exports.createErrorLogger = createErrorLogger;
function createLoggers(...namespaces) {
    return {
        info: createInfoLogger(...namespaces),
        debug: createDebugLogger(...namespaces),
        warn: createWarnLogger(...namespaces),
        error: createErrorLogger(...namespaces),
    };
}
exports.createLoggers = createLoggers;
