// @node-wot/core/dist/protocol-listener-registry.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const content_serdes_1 = __importDefault(require("./content-serdes"));
class ProtocolListenerRegistry {
    constructor() {
        this.listeners = new Map();
    }
    register(affordance, formIndex, listener) {
        if (!affordance.forms[formIndex]) {
            throw new Error("Can't register the listener for affordance with formIndex. The affordance does not contain the form");
        }
        let formMap = this.listeners.get(affordance);
        if (!formMap) {
            formMap = new Map();
            this.listeners.set(affordance, formMap);
        }
        let listeners = formMap.get(formIndex);
        if (!listeners) {
            listeners = [];
            formMap.set(formIndex, listeners);
        }
        listeners.push(listener);
    }
    unregister(affordance, formIndex, listener) {
        const formMap = this.listeners.get(affordance);
        if (!formMap) {
            throw new Error("Not found");
        }
        const listeners = formMap.get(formIndex);
        if (!listeners) {
            throw new Error("Form not found");
        }
        const index = listeners.indexOf(listener);
        if (index < 0) {
            throw new Error("Form not found");
        }
        listeners.splice(index, 1);
    }
    unregisterAll() {
        this.listeners.clear();
    }
    notify(affordance, data, schema, formIndex) {
        var _a;
        const formMap = (_a = this.listeners.get(affordance)) !== null && _a !== void 0 ? _a : ProtocolListenerRegistry.EMPTY_MAP;
        if (formIndex !== undefined) {
            const listeners = formMap.get(formIndex);
            if (listeners) {
                const contentType = affordance.forms[formIndex].contentType;
                const content = content_serdes_1.default.valueToContent(data, schema, contentType);
                listeners.forEach((listener) => listener(content));
                return;
            }
        }
        for (const [index, value] of formMap) {
            const contentType = affordance.forms[index].contentType;
            const content = content_serdes_1.default.valueToContent(data, schema, contentType);
            value.forEach((listener) => listener(content));
        }
    }
}
exports.default = ProtocolListenerRegistry;
ProtocolListenerRegistry.EMPTY_MAP = new Map();
