// @node-wot/binding-http/dist/credential.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TuyaCustomBearer = exports.OAuthCredential = exports.BasicKeyCredential = exports.BearerCredential = exports.BasicCredential = exports.Credential = void 0;
const node_fetch_1 = __importStar(require("node-fetch"));
const crypto = __importStar(require("crypto"));
const queryString = __importStar(require("query-string"));
class Credential {
}
exports.Credential = Credential;
class BasicCredential extends Credential {
    constructor({ username, password }, options) {
        super();
        if (username === undefined || password === undefined || username === null || password === null) {
            throw new Error(`No Basic credentials for Thing`);
        }
        this.username = username;
        this.password = password;
        this.options = options;
    }
    sign(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = request.clone();
            let headerName = "authorization";
            if (this.options !== undefined && this.options.in === "header" && this.options.name !== undefined) {
                headerName = this.options.name;
            }
            result.headers.set(headerName, "Basic " + Buffer.from(this.username + ":" + this.password).toString("base64"));
            return result;
        });
    }
}
exports.BasicCredential = BasicCredential;
class BearerCredential extends Credential {
    constructor({ token }, options) {
        super();
        if (token === undefined || token === null) {
            throw new Error(`No Bearer credentials for Thing`);
        }
        this.token = token;
        this.options = options;
    }
    sign(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = request.clone();
            let headerName = "authorization";
            if (this.options.in === "header" && this.options.name !== undefined) {
                headerName = this.options.name;
            }
            result.headers.set(headerName, "Bearer " + this.token);
            return result;
        });
    }
}
exports.BearerCredential = BearerCredential;
class BasicKeyCredential extends Credential {
    constructor({ apiKey }, options) {
        super();
        if (apiKey === undefined || apiKey === null) {
            throw new Error(`No API key credentials for Thing`);
        }
        this.apiKey = apiKey;
        this.options = options;
    }
    sign(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = request.clone();
            let headerName = "authorization";
            if (this.options.in === "header" && this.options.name !== undefined) {
                headerName = this.options.name;
            }
            result.headers.append(headerName, this.apiKey);
            return result;
        });
    }
}
exports.BasicKeyCredential = BasicKeyCredential;
class OAuthCredential extends Credential {
    constructor(token, refresh) {
        super();
        this.token = token;
        this.refresh = refresh;
        this.token = token;
    }
    sign(request) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.token instanceof Promise) {
                const tokenRequest = this.token;
                this.token = yield tokenRequest;
            }
            let tempRequest = { url: request.url, headers: {} };
            tempRequest = this.token.sign(tempRequest);
            const mergeHeaders = new node_fetch_1.Request(request, tempRequest);
            return mergeHeaders;
        });
    }
    refreshToken() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.token instanceof Promise) {
                throw new Error("Uninitialized token. You have to call sing before refresh");
            }
            let newToken;
            if (this.refresh) {
                newToken = yield this.refresh();
            }
            else {
                newToken = yield this.token.refresh();
            }
            return new OAuthCredential(newToken, this.refresh);
        });
    }
}
exports.OAuthCredential = OAuthCredential;
class TuyaCustomBearer extends Credential {
    constructor(credentials, scheme) {
        super();
        this.key = credentials.key;
        this.secret = credentials.secret;
        this.baseUri = scheme.baseUri;
    }
    sign(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const isTokenExpired = this.isTokenExpired();
            if (this.token === undefined || this.token === "" || isTokenExpired)
                yield this.requestAndRefreshToken(isTokenExpired);
            const url = request.url;
            const body = request.body ? request.body.read().toString() : "";
            const headers = this.getHeaders(true, request.headers.raw(), body, url, request.method);
            Object.assign(headers, request.headers.raw());
            return new node_fetch_1.Request(url, { method: request.method, body: body !== "" ? body : null, headers: headers });
        });
    }
    requestAndRefreshToken(refresh) {
        return __awaiter(this, void 0, void 0, function* () {
            const headers = this.getHeaders(false, {}, "", null, null);
            const request = {
                headers: headers,
                method: "GET",
            };
            let url = `${this.baseUri}/token?grant_type=1`;
            if (refresh) {
                url = `${this.baseUri}/token/${this.refreshToken}`;
            }
            const data = yield (yield (0, node_fetch_1.default)(url, request)).json();
            if (data.success) {
                this.token = data.result.access_token;
                this.refreshToken = data.result.refresh_token;
                this.expireTime = new Date(Date.now() + data.result.expire_time * 1000);
            }
            else {
                throw new Error("token fetch failed");
            }
        });
    }
    getHeaders(NormalRequest, headers, body, url, method) {
        const requestTime = Date.now().toString();
        const replaceUri = this.baseUri.replace("/v1.0", "");
        const _url = url ? url.replace(`${replaceUri}`, "") : null;
        const sign = this.requestSign(NormalRequest, requestTime, body, headers, _url, method);
        return {
            t: requestTime,
            client_id: this.key,
            sign_method: "HMAC-SHA256",
            sign,
            access_token: this.token || "",
        };
    }
    requestSign(NormalRequest, requestTime, body, headers, path, method) {
        const bodyHash = crypto.createHash("sha256").update(body).digest("hex");
        let signUrl = "/v1.0/token?grant_type=1";
        const headerString = "";
        let useToken = "";
        const _method = method || "GET";
        if (NormalRequest) {
            useToken = this.token;
            const pathQuery = queryString.parse(path.split("?")[1]);
            let query = {};
            query = Object.assign(query, pathQuery);
            const sortedQuery = {};
            Object.keys(query)
                .sort()
                .forEach((i) => {
                sortedQuery[i] = query[i];
            });
            const qs = queryString.stringify(sortedQuery);
            signUrl = decodeURIComponent(qs ? `${path.split("?")[0]}?${qs}` : path);
        }
        const endStr = [this.key, useToken, requestTime, [_method, bodyHash, headerString, signUrl].join("\n")].join("");
        const sign = crypto.createHmac("sha256", this.secret).update(endStr).digest("hex").toUpperCase();
        return sign;
    }
    isTokenExpired() {
        return this.expireTime ? Date.now() > this.expireTime.getTime() : false;
    }
}
exports.TuyaCustomBearer = TuyaCustomBearer;
