// @node-wot/core/dist/exposed-thing.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Subject_1 = require("rxjs/Subject");
const TD = __importStar(require("@node-wot/td-tools"));
const helpers_1 = __importDefault(require("./helpers"));
const interaction_output_1 = require("./interaction-output");
const stream_1 = require("stream");
const protocol_helpers_1 = __importDefault(require("./protocol-helpers"));
const es2018_1 = require("web-streams-polyfill/ponyfill/es2018");
const content_serdes_1 = __importDefault(require("./content-serdes"));
const protocol_listener_registry_1 = __importDefault(require("./protocol-listener-registry"));
const logger_1 = require("./logger");
const { debug } = (0, logger_1.createLoggers)("core", "exposed-thing");
class ExposedThing extends TD.Thing {
    constructor(servient, thingModel = {}) {
        var _a, _b;
        super();
        this.__propertyHandlers = new Map();
        this.__actionHandlers = new Map();
        this.__eventHandlers = new Map();
        this.__propertyListeners = new protocol_listener_registry_1.default();
        this.__eventListeners = new protocol_listener_registry_1.default();
        this.getServient = () => {
            return servient;
        };
        this.getSubjectTD = new (class {
            constructor() {
                this.subjectTDChange = new Subject_1.Subject();
                this.getSubject = () => {
                    return this.subjectTDChange;
                };
            }
        })().getSubject;
        this.id = (_a = thingModel.id) !== null && _a !== void 0 ? _a : "";
        this.title = (_b = thingModel.title) !== null && _b !== void 0 ? _b : "";
        this.security = "";
        this.securityDefinitions = {};
        this.properties = {};
        this.actions = {};
        this.events = {};
        const clonedModel = JSON.parse(JSON.stringify(thingModel));
        Object.assign(this, clonedModel);
        TD.setContextLanguage(this, TD.DEFAULT_CONTEXT_LANGUAGE, false);
    }
    getThingDescription() {
        return JSON.parse(TD.serializeTD(this), (key, value) => {
            if (key === "__propertyHandlers" ||
                key === "__actionHandlers" ||
                key === "__eventHandlers" ||
                key === "__propertyListeners" ||
                key === "__eventListeners") {
                return undefined;
            }
            return value;
        });
    }
    emitEvent(name, data) {
        if (this.events[name]) {
            const eventAffordance = this.events[name];
            this.__eventListeners.notify(eventAffordance, data, eventAffordance.data);
        }
        else {
            throw new Error("NotFoundError for event '" + name + "'");
        }
    }
    emitPropertyChange(name) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (this.properties[name]) {
                const property = this.properties[name];
                const readHandler = (_a = this.__propertyHandlers.get(name)) === null || _a === void 0 ? void 0 : _a.readHandler;
                if (!readHandler) {
                    throw new Error("Can't read property readHandler is not defined. Did you forget to register a readHandler?");
                }
                const data = yield readHandler();
                this.__propertyListeners.notify(property, data, property);
            }
            else {
                throw new Error("NotFoundError for property '" + name + "'");
            }
        });
    }
    expose() {
        debug(`ExposedThing '${this.title}' exposing all Interactions and TD`);
        return new Promise((resolve, reject) => {
            this.getServient()
                .expose(this)
                .then(() => {
                this.getSubjectTD().next(this.getThingDescription());
                resolve();
            })
                .catch((err) => reject(err));
        });
    }
    destroy() {
        return __awaiter(this, void 0, void 0, function* () {
            debug(`ExposedThing '${this.title}' destroying the thing and its interactions`);
            yield this.getServient().destroyThing(this.id);
            this.__eventListeners.unregisterAll();
            this.__propertyListeners.unregisterAll();
            this.__eventHandlers.clear();
            this.__propertyHandlers.clear();
            this.__eventHandlers.clear();
            this.getSubjectTD().next();
        });
    }
    setPropertyReadHandler(propertyName, handler) {
        debug(`ExposedThing '${this.title}' setting read handler for '${propertyName}'`);
        if (this.properties[propertyName]) {
            if (this.properties[propertyName].writeOnly) {
                throw new Error(`ExposedThing '${this.title}' cannot set read handler for property '${propertyName}' due to writeOnly flag`);
            }
            else {
                let propertyHandler = this.__propertyHandlers.get(propertyName);
                if (propertyHandler) {
                    propertyHandler.readHandler = handler;
                }
                else {
                    propertyHandler = { readHandler: handler };
                }
                this.__propertyHandlers.set(propertyName, propertyHandler);
            }
        }
        else {
            throw new Error(`ExposedThing '${this.title}' has no Property '${propertyName}'`);
        }
        return this;
    }
    setPropertyWriteHandler(propertyName, handler) {
        debug(`ExposedThing '${this.title}' setting write handler for '${propertyName}'`);
        if (this.properties[propertyName]) {
            if (this.properties[propertyName].readOnly) {
                throw new Error(`ExposedThing '${this.title}' cannot set write handler for property '${propertyName}' due to readOnly flag`);
            }
            else {
                let propertyHandler = this.__propertyHandlers.get(propertyName);
                if (propertyHandler) {
                    propertyHandler.writeHandler = handler;
                }
                else {
                    propertyHandler = { writeHandler: handler };
                }
                this.__propertyHandlers.set(propertyName, propertyHandler);
            }
        }
        else {
            throw new Error(`ExposedThing '${this.title}' has no Property '${propertyName}'`);
        }
        return this;
    }
    setPropertyObserveHandler(name, handler) {
        debug(`ExposedThing '${this.title}' setting property observe handler for '${name}'`);
        if (this.properties[name]) {
            if (!this.properties[name].observable) {
                throw new Error(`ExposedThing '${this.title}' cannot set observe handler for property '${name}' since the observable flag is set to false`);
            }
            else {
                let propertyHandler = this.__propertyHandlers.get(name);
                if (propertyHandler) {
                    propertyHandler.observeHandler = handler;
                }
                else {
                    propertyHandler = { observeHandler: handler };
                }
                this.__propertyHandlers.set(name, propertyHandler);
            }
        }
        else {
            throw new Error(`ExposedThing '${this.title}' has no Property '${name}'`);
        }
        return this;
    }
    setPropertyUnobserveHandler(name, handler) {
        debug(`ExposedThing '${this.title}' setting property unobserve handler for '${name}'`);
        if (this.properties[name]) {
            if (!this.properties[name].observable) {
                throw new Error(`ExposedThing '${this.title}' cannot set unobserve handler for property '${name}' due to missing observable flag`);
            }
            else {
                let propertyHandler = this.__propertyHandlers.get(name);
                if (propertyHandler) {
                    propertyHandler.unobserveHandler = handler;
                }
                else {
                    propertyHandler = { unobserveHandler: handler };
                }
                this.__propertyHandlers.set(name, propertyHandler);
            }
        }
        else {
            throw new Error(`ExposedThing '${this.title}' has no Property '${name}'`);
        }
        return this;
    }
    setActionHandler(actionName, handler) {
        debug(`ExposedThing '${this.title}' setting action handler for '${actionName}'`);
        if (this.actions[actionName]) {
            this.__actionHandlers.set(actionName, handler);
        }
        else {
            throw new Error(`ExposedThing '${this.title}' has no Action '${actionName}'`);
        }
        return this;
    }
    setEventSubscribeHandler(name, handler) {
        debug(`ExposedThing '${this.title}' setting event subscribe handler for '${name}'`);
        if (this.events[name]) {
            let eventHandler = this.__eventHandlers.get(name);
            if (eventHandler) {
                eventHandler.subscribe = handler;
            }
            else {
                eventHandler = { subscribe: handler };
            }
            this.__eventHandlers.set(name, eventHandler);
        }
        else {
            throw new Error(`ExposedThing '${this.title}' has no Event '${name}'`);
        }
        return this;
    }
    setEventUnsubscribeHandler(name, handler) {
        debug(`ExposedThing '${this.title}' setting event unsubscribe handler for '${name}'`);
        if (this.events[name]) {
            let eventHandler = this.__eventHandlers.get(name);
            if (eventHandler) {
                eventHandler.unsubscribe = handler;
            }
            else {
                eventHandler = { unsubscribe: handler };
            }
            this.__eventHandlers.set(name, eventHandler);
        }
        else {
            throw new Error(`ExposedThing '${this.title}' has no Event '${name}'`);
        }
        return this;
    }
    setEventHandler(name, handler) {
        debug(`ExposedThing '${this.title}' setting event handler for '${name}'`);
        if (this.events[name]) {
            let eventHandler = this.__eventHandlers.get(name);
            if (eventHandler) {
                eventHandler.handler = handler;
            }
            else {
                eventHandler = { handler: handler };
            }
            this.__eventHandlers.set(name, eventHandler);
        }
        else {
            throw new Error(`ExposedThing '${this.title}' has no Event '${name}'`);
        }
        return this;
    }
    handleInvokeAction(name, inputContent, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.actions[name]) {
                debug(`ExposedThing '${this.title}' has Action state of '${name}'`);
                const handler = this.__actionHandlers.get(name);
                if (handler != null) {
                    debug(`ExposedThing '${this.title}' calls registered handler for Action '${name}'`);
                    helpers_1.default.validateInteractionOptions(this, this.actions[name], options);
                    const form = this.actions[name].forms
                        ? this.actions[name].forms[options.formIndex]
                        : { contentType: "application/json" };
                    const result = yield handler(new interaction_output_1.InteractionOutput(inputContent, form, this.actions[name].input), options);
                    if (result) {
                        return content_serdes_1.default.valueToContent(result, this.actions[name].output, form.contentType);
                    }
                }
                else {
                    throw new Error(`ExposedThing '${this.title}' has no handler for Action '${name}'`);
                }
            }
            else {
                throw new Error(`ExposedThing '${this.title}', no action found for '${name}'`);
            }
        });
    }
    handleReadProperty(propertyName, options) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            if (this.properties[propertyName]) {
                debug(`ExposedThing '${this.title}' has Action state of '${propertyName}'`);
                const readHandler = (_a = this.__propertyHandlers.get(propertyName)) === null || _a === void 0 ? void 0 : _a.readHandler;
                if (readHandler != null) {
                    debug(`ExposedThing '${this.title}' calls registered readHandler for Property '${propertyName}'`);
                    helpers_1.default.validateInteractionOptions(this, this.properties[propertyName], options);
                    const result = yield readHandler(options);
                    const form = this.properties[propertyName].forms
                        ? this.properties[propertyName].forms[options.formIndex]
                        : { contentType: "application/json" };
                    return content_serdes_1.default.valueToContent(result, this.properties[propertyName], (_b = form === null || form === void 0 ? void 0 : form.contentType) !== null && _b !== void 0 ? _b : "application/json");
                }
                else {
                    throw new Error(`ExposedThing '${this.title}' has no readHandler for Property '${propertyName}'`);
                }
            }
            else {
                throw new Error(`ExposedThing '${this.title}', no property found for '${propertyName}'`);
            }
        });
    }
    _handleReadProperties(propertyNames, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const promises = [];
            for (const propertyName of propertyNames) {
                const form = this.properties[propertyName].forms.find((form) => form.contentType === "application/json" || !form.contentType);
                if (!form) {
                    continue;
                }
                promises.push(this.handleReadProperty(propertyName, options));
            }
            try {
                const output = new Map();
                const results = yield Promise.all(promises);
                for (let i = 0; i < results.length; i++) {
                    output.set(propertyNames[i], results[i]);
                }
                return output;
            }
            catch (error) {
                throw new Error(`ConsumedThing '${this.title}', failed to read properties: ${propertyNames}.\n Error: ${error}`);
            }
        });
    }
    handleReadAllProperties(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const propertyNames = [];
            for (const propertyName in this.properties) {
                propertyNames.push(propertyName);
            }
            return yield this._handleReadProperties(propertyNames, options);
        });
    }
    handleReadMultipleProperties(propertyNames, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._handleReadProperties(propertyNames, options);
        });
    }
    handleWriteProperty(propertyName, inputContent, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (this.properties[propertyName]) {
                if (this.properties[propertyName].readOnly && this.properties[propertyName].readOnly === true) {
                    throw new Error(`ExposedThing '${this.title}', property '${propertyName}' is readOnly`);
                }
                helpers_1.default.validateInteractionOptions(this, this.properties[propertyName], options);
                const writeHandler = (_a = this.__propertyHandlers.get(propertyName)) === null || _a === void 0 ? void 0 : _a.writeHandler;
                const form = this.properties[propertyName].forms
                    ? this.properties[propertyName].forms[options.formIndex]
                    : {};
                if (writeHandler != null) {
                    yield writeHandler(new interaction_output_1.InteractionOutput(inputContent, form, this.properties[propertyName]), options);
                }
                else {
                    throw new Error(`ExposedThing '${this.title}' has no writeHandler for Property '${propertyName}'`);
                }
            }
            else {
                throw new Error(`ExposedThing '${this.title}', no property found for '${propertyName}'`);
            }
        });
    }
    handleWriteMultipleProperties(valueMap, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const promises = [];
            for (const propertyName in valueMap) {
                const form = this.properties[propertyName].forms.find((form) => form.contentType === "application/json" || !form.contentType);
                if (!form) {
                    continue;
                }
                promises.push(this.handleWriteProperty(propertyName, valueMap.get(propertyName), options));
            }
            try {
                yield Promise.all(promises);
            }
            catch (error) {
                throw new Error(`ExposedThing '${this.title}', failed to write multiple properties. ${error.message}`);
            }
        });
    }
    handleSubscribeEvent(name, listener, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (this.events[name]) {
                helpers_1.default.validateInteractionOptions(this, this.events[name], options);
                const formIndex = protocol_helpers_1.default.getFormIndexForOperation(this.events[name], "event", "subscribeevent", options.formIndex);
                if (formIndex !== -1) {
                    this.__eventListeners.register(this.events[name], formIndex, listener);
                    debug(`ExposedThing '${this.title}' subscribes to event '${name}'`);
                }
                else {
                    throw new Error(`ExposedThing '${this.title}', no property listener from found for '${name}' with form index '${options.formIndex}'`);
                }
                const subscribe = (_a = this.__eventHandlers.get(name)) === null || _a === void 0 ? void 0 : _a.subscribe;
                if (subscribe) {
                    yield subscribe(options);
                }
                debug(`ExposedThing '${this.title}' subscribes to event '${name}'`);
            }
            else {
                throw new Error(`ExposedThing '${this.title}', no event found for '${name}'`);
            }
        });
    }
    handleUnsubscribeEvent(name, listener, options) {
        var _a;
        if (this.events[name]) {
            helpers_1.default.validateInteractionOptions(this, this.events[name], options);
            const formIndex = protocol_helpers_1.default.getFormIndexForOperation(this.events[name], "event", "unsubscribeevent", options.formIndex);
            if (formIndex !== -1) {
                this.__eventListeners.unregister(this.events[name], formIndex, listener);
            }
            else {
                throw new Error(`ExposedThing '${this.title}', no event listener from found for '${name}' with form index '${options.formIndex}'`);
            }
            const unsubscribe = (_a = this.__eventHandlers.get(name)) === null || _a === void 0 ? void 0 : _a.unsubscribe;
            if (unsubscribe) {
                unsubscribe(options);
            }
            debug(`ExposedThing '${this.title}' unsubscribes from event '${name}'`);
        }
        else {
            throw new Error(`ExposedThing '${this.title}', no event found for '${name}'`);
        }
    }
    handleObserveProperty(name, listener, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (this.properties[name]) {
                helpers_1.default.validateInteractionOptions(this, this.properties[name], options);
                const formIndex = protocol_helpers_1.default.getFormIndexForOperation(this.properties[name], "property", "observeproperty", options.formIndex);
                if (formIndex !== -1) {
                    this.__propertyListeners.register(this.properties[name], formIndex, listener);
                    debug(`ExposedThing '${this.title}' subscribes to property '${name}'`);
                }
                else {
                    throw new Error(`ExposedThing '${this.title}', no property listener from found for '${name}' with form index '${options.formIndex}'`);
                }
                const observeHandler = (_a = this.__propertyHandlers.get(name)) === null || _a === void 0 ? void 0 : _a.observeHandler;
                if (observeHandler) {
                    yield observeHandler(options);
                }
            }
            else {
                throw new Error(`ExposedThing '${this.title}', no property found for '${name}'`);
            }
        });
    }
    handleUnobserveProperty(name, listener, options) {
        var _a;
        if (this.properties[name]) {
            helpers_1.default.validateInteractionOptions(this, this.properties[name], options);
            const formIndex = protocol_helpers_1.default.getFormIndexForOperation(this.properties[name], "property", "unobserveproperty", options.formIndex);
            if (formIndex !== -1) {
                this.__propertyListeners.unregister(this.properties[name], formIndex, listener);
            }
            else {
                throw new Error(`ExposedThing '${this.title}', no property listener from found for '${name}' with form index '${options.formIndex}'`);
            }
            const unobserveHandler = (_a = this.__propertyHandlers.get(name)) === null || _a === void 0 ? void 0 : _a.unobserveHandler;
            if (unobserveHandler) {
                unobserveHandler(options);
            }
        }
        else {
            throw new Error(`ExposedThing '${this.title}', no property found for '${name}'`);
        }
    }
    static interactionInputToReadable(input) {
        let body;
        if (typeof ReadableStream !== "undefined" && input instanceof ReadableStream) {
            body = protocol_helpers_1.default.toNodeStream(input);
        }
        else if (input instanceof es2018_1.ReadableStream) {
            body = protocol_helpers_1.default.toNodeStream(input);
        }
        else if (Array.isArray(input) || typeof input === "object") {
            body = stream_1.Readable.from(Buffer.from(JSON.stringify(input), "utf-8"));
        }
        else {
            body = stream_1.Readable.from(Buffer.from(input.toString(), "utf-8"));
        }
        return body;
    }
}
exports.default = ExposedThing;
